protocols/consensus-gossip/tuple-space.md

```diff
@@ -0,0 +1,192 @@
+
+In the case of the Tendermint algorithm we note that staleness comes from adding newer messages (belonging to higher rounds and heights) to the tuple space.
+Hence, in Approach Two, if as an optimization these newer messages are exchanged first, then the stale messages can be excluded before being shared to other nodes that might have forgotten them and tombstones may not be needed at all.
+
+
+### Querying the Tuple Space
+
+The tuple space is consulted through queries, which have the same form as the entries.
+Queries return all entries whose values match those in the query, where a `*` matches all values.
+For example, suppose a node's local view of the tuple space has the following entries, here organized as rows of a table for easier visualization:
+
+| Height | Round | Step     | Validator | Value |
+|--------|-------|----------|-----------|-------|
+| 1      | 0     | Proposal | p         | pval  |
+| 1      | 0     | PreVote  | p         | vval  |
+| 1      | 1     | PreCommit| q         | cval  |
+| 2      | 0     | Proposal | p         | pval' |
+| 2      | 2     | PreVote  | q         | vval' |
+| 2      | 3     | PreCommit| q         | cval' |
+
+- Query $\lang 0, 0, Proposal, p, * \rang$ returns $\{ \lang 0, 0, Proposal, p, pval \rang \}$
+- Query $\lang 0, 0, *, p, * \rang$ returns $\{ \lang 0, 0, Proposal, p, pval \rang,  \lang 0, 0, PreVote, p, vval \rang \}$.
+
+
+#### Local views
+
+Let $T$ be the tuple space and $t_p$ be node $p$'s approximation of $T$, also denominated $p$'s **local views**.
+Because of the asynchronous nature of distributed systems, local views may not include entries in the space or may still include entries no longer in the space.
+
+Formally, let $P$ be the set of validators; $T = \cup_P t_p$ and $e \in T \Leftrightarrow \exists p \in P, e \in t_p \land \not\exists q \in Q, \bar{e}\in t_q$
+
+Nodes can only query their own local views, not $T$.
+If needed for disambiguation, queries are subscripted with the node being queried.
+
+#### State Validity
+
+Let $\text{ValSet}_h$ be the set of validators of height $h$ and $\text{Prop}_{h,r}$ be the proposal of round $r$ of height $h$.
+
+Given that each validator can execute each step only once per round, a query that specifies height, round, step and validator must either return empty or a single tuple.
+
+- $\forall h \in \N, r \in \N, s \in \{\text{Proposal, PreVote, PreCommit}\}, v \in \text{ValSet}_h$,  $\lang h, r, s, v, * \rang$ returns at most one value.
+
+In the specific case of the Proposal step, only the proposer of the round can have a matching entry.
+
+- $\forall h \in \N, r \in \N, \lang h, r, \text{Proposal}, *, * \rang$ returns at most one value and it also matches $\lang h, r, \text{Proposal}, \text{Prop}_{h,r}, * \rang$.
+
+A violation of these rules is a proof of misbehavior.
```

Comment:
Currently, we don't consider double proposing as a misbehavior. We only detect and punish double-voting.